### DISCUSSION TITLE
Can ESP32 support WIFI, BT, BLE, WSS, MQTT and decorder at the same time?

### DISCUSSION THREAD

**Discussion by Sakuranda** [espressif/arduino-esp32#10285] [Q&A]
Hi,
ESP32 only has 320kb RAM. Although we have 8M PSRAM, but it's still leaked of memory.

There's our functions to be achieved:
1. Running WIFI STA to commuicate with phone, BLE to commuicate and locatilze with other esp32, BT audio a2dp to play music with another esp32.
2. Using WebSockets to authorize the server to retrieve data, using MQTT or other protocol to transmit and receive data from server.
3. Record and decord mp3 audio between server and esp32's mic.

Can esp32 support to run these functions within the same project? We're developed with platformio+arduino framework, but if it's enough for chip resource in theory, we'll change to esp-idf.

I also asked in esp32 forum https://www.esp32.com/viewtopic.php?f=2&t=41682.

Thanks!

**Comment by Jason2866** [comment]
Maybe https://github.com/pschatzmann/ESP32-A2DP

**Reply by Sakuranda** [reply]
Hi Jason,
Thanks for your reply. I saw this useful lib but I also want to know can ESP32 support these tons of functions at the same time? If it's impossible, we will make another choice.

Best regards,
Sakuranda

**Reply by Jason2866** [reply]
Depending on your code ;-). In general it should be possible to do.

**Reply by Sakuranda** [reply]
Hi,
Here's our demo code [esp32_test](https://github.com/Sakuranda/esp32_test), we're just run the a2dp function and start websockets init, it will be crashed at wifi init.

There's the log:
```
cfg end
Free PSRAM after Wi-Fi init: 4172099 bytes
Free heap after Wi-Fi init: 48296 bytes
[D] AudioIO.h : 30 - void audio_tools::TransformationReader<T>::begin(T*, Stream*) [with T = audio_tools::ReformatBaseStream]
[D] AudioEncoded.h : 140 - virtual bool audio_tools::EncodedAudioOutput::begin()
[D] AudioEncoded.h : 142 - virtual bool audio_tools::EncodedAudioOutput::begin()
[D] AudioCodecsBase.h : 27 - virtual void audio_tools::AudioDecoder::setAudioInfo(audio_tools::AudioInfo)
[D] CodecMP3Helix.h : 82 - virtual bool audio_tools::MP3DecoderHelix::begin()
[D] MetaDataFilter.h : 30 - void audio_tools::MetaDataFilter<Decoder>::begin() [with Decoder = libhelix::MP3DecoderHelix]
dec end
Free PSRAM after Wi-Fi init: 4140955 bytes
Free heap after Wi-Fi init: 48296 bytes
E (9152) wifi:Expected to init 4 rx buffer, actual is 2
E (9154) wifi_init: Failed to deinit Wi-Fi driver (0x3001)
E (9155) wifi_init: Failed to deinit Wi-Fi (0x3001)
[  4334][E][WiFiGeneric.cpp:685] wifiLowLevelInit(): esp_wifi_init 257
[  4352][E][WiFiSTA.cpp:227] begin(): STA enable failed!
Free PSRAM after Wi-Fi init: 4140955 bytes
Free heap after Wi-Fi init: 28216 bytes
Connecting to WiFi...
Connecting to WiFi...
```

So we're doubt esp32's ability, hope to your reply. Thanks!

Best regards,
Sakuranda
